### PR TITLE
Add docs build job

### PR DESCRIPTION
**Pull request by AyodeAwe**
The PR adds a `docs_build` process to the PR and Build workflows for this repository. The generated docs are synced to s3 for only the build workflows.

### CODE REVIEW

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:32]:
```suggestion
rapids-logger "Build Doxygen docs"
```

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:38]:
```suggestion
rapids-logger "Build Sphinx docs"
```

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:15]:
```suggestion
rapids-mamba-retry env create --force -f env.yaml -n docs
conda activate docs
```

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:41]:
```suggestion
sphinx-build -b dirhtml source _html -W
sphinx-build -b text source _text -W
```

@AjayThorve added this flag in https://github.com/rapidsai/cuml/pull/5155/files, which I think is a good idea to add to all of the repositories.

It will treat warnings as errors as indicated by the sphinx docs: https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-W

If any errors appear as a result of this, we can try to resolve them ourselves or solicit feedback from the cuspatial team.

File: dependencies.yaml
Comment (ajschmidt8) [inline dependencies.yaml:148]:
```suggestion
          - doxygen
          - ipython
          - myst-parser
          - nbsphinx
          - numpydoc
          - pydata-sphinx-theme
```

Let's alphabetize these to keep things neat.

FWIW, I don't alphabetize these by hand.

If you highlight the lines in VS Code, you can use the `Sort Lines Ascending/Descending` commands in the [command palette](https://docs.github.com/en/codespaces/codespaces-reference/using-the-vs-code-command-palette-in-codespaces#accessing-the-vs-code-command-palette).

File: dependencies.yaml
Comment (ajschmidt8) [inline dependencies.yaml:37]:
```suggestion
      - cudatoolkit
      - docs_build
      - py_version
```
alphabetize

**Review by ajschmidt8 — COMMENTED** [review]
Looks good. Added a few comments

**Comment by AyodeAwe** [comment]
```
Warning, treated as error:
/opt/conda/envs/docs/lib/python3.9/site-packages/cuspatial/core/geoseries.py:docstring of cuspatial.core.geoseries.GeoSeries.contains_properly:39:Unexpected indentation.
```

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:31]:
```suggestion
```

File: ci/build_docs.sh
Comment (ajschmidt8) [inline ci/build_docs.sh:37]:
```suggestion
```

File: dependencies.yaml
Comment (ajschmidt8) [inline dependencies.yaml:148]: `sphinx` should probably be added explicitly here since we use `sphinx-build` in the scripts.

It's probably being pulled in transitively right now.

File: dependencies.yaml
Comment (ajschmidt8) [inline dependencies.yaml:36]: This comment is relevant: https://github.com/rapidsai/rmm/pull/1189#discussion_r1073685857

File: dependencies.yaml
Comment (ajschmidt8) [inline dependencies.yaml:139]: https://github.com/rapidsai/rmm/pull/1189#discussion_r1073685857

**Review by thomcom — COMMENTED** [review]
Are these inline pydocs changes purposeful? I ask because it seems like there would be a lot more changes across the codebase to match the new formatting requirement, if there is one.

File: .github/workflows/build.yaml
Comment (ajschmidt8) [inline .github/workflows/build.yaml:59]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

File: .github/workflows/pr.yaml
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:53]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

**Comment by ajschmidt8** [comment]
> Are these inline pydocs changes purposeful? I ask because it seems like there would be a lot more changes across the codebase to match the new formatting requirement, if there is one.

@thomcom, we've enabled the `-W` flag for the Sphinx builds, which treats warnings as errors (e.g. [these](https://github.com/rapidsai/cuspatial/actions/runs/4075479030/jobs/7022553917#step:6:1617)).

It seems like there are only 11 warnings and they all look whitespace related, so it shouldn't be too hard for us to address them in this PR.

If it gets out of hand, we can address them in a follow-up PR.

**Comment by ajschmidt8** [comment]
I've just pushed a few additional changes and some reversions.

I think this is almost the minimum set of changes needed to build the docs without any warnings.

I'm expecting the latest commit as of this writing (4f60b18) to throw a `duplicate label` error from the [`autosectionlabel` plugin](https://www.sphinx-doc.org/en/master/usage/extensions/autosectionlabel.html).

Once that happens, I will push the change below to suppress it since it's a relatively benign warning.

```py
suppress_warnings = [
    "autosectionlabel.*"
]
```

It stems from the fact that the https://docs.rapids.ai/api/cuspatial/nightly/user_guide/cuspatial_api_examples.html page has two section headers called `cuspatial.polygon_bounding_boxes`.

As a result of this, the first occurrence gets a correct anchor attached, but the second occurrence gets an anchor of `#id1`. See here:

- https://docs.rapids.ai/api/cuspatial/nightly/user_guide/cuspatial_api_examples.html#cuspatial.polygon_bounding_boxes
- https://docs.rapids.ai/api/cuspatial/nightly/user_guide/cuspatial_api_examples.html#id1

The consequence of this is that you can't confidently link to either section using the `:ref:` syntax outlined in the [autosectionlabel docs](https://www.sphinx-doc.org/en/master/usage/extensions/autosectionlabel.html) since there is a duplicately named section.

However, I didn't see any instances of `:ref:` being used anywhere.

@thomcom, I'll defer to you to determine if you want to leave the `suppress_warnings` entry or properly address the warning in a follow-up PR.

**Comment by thomcom** [comment]
I'd be interested in addressing the warnings in a follow up PR, we can chat about how to get my pre-commit set up to give me docs warnings in addition to the normal commit errors. Given that `cuspatial.polygon_bounding_boxes` is already defined in its own section, I'm ok with dropping that from the list of three functions under `Indexed Spatial Joins`.

**Review by ajschmidt8 — APPROVED** [review]

**Comment by ajschmidt8** [comment]
@thomcom, I just pushed two changes based on your reply.

Can you review / approve this PR when you have a moment? It should be good to go now.

**Review by thomcom — APPROVED** [review]

**Pull request merged**
